### PR TITLE
#24 Feature: New offline dictionary implementation without need to pad bytes elements (variable length values)

### PR DESCRIPTION
**Pull request by buchireddy**
See #24 for more details.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java [hunk 1]
```diff
@@ -154,24 +155,18 @@ public void build()
...
           _stringValueToIndexMap.put(value, i);
-          byte[] valueBytes = StringUtil.encodeUtf8(value);
-          sortedStringBytes[i] = valueBytes;
-          _numBytesPerEntry = Math.max(_numBytesPerEntry, valueBytes.length);
         }
 
         // Backward-compatible: index file is always big-endian
-        try (PinotDataBuffer dataBuffer = PinotDataBuffer
-            .mapFile(_dictionaryFile, false, 0, (long) numValues * _numBytesPerEntry, ByteOrder.BIG_ENDIAN,
+        long size = VarLengthStringsReaderWriter.getRequiredSize(sortedStrings);
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java]: we need to support both modes.  We need the new implementation for byte and string column and only when the standard deviation is high.

Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java]: Are you saying we calculate the stdDev and decide to use whether fixed or var length dict based on the std dev on-the-fly?

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+ * under the License.
+ */
+package org.apache.pinot.core.io.util;
+
+import com.google.common.base.Preconditions;
+import java.io.Closeable;
+import java.io.IOException;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+public class VarLengthBytesValueReaderWriter implements Closeable, ValueReader {
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:48]: we just need reader

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+ */
+public class VarLengthBytesValueReaderWriter implements Closeable, ValueReader {
+
+  /**
+   * Header used to identify the dictionary files written in variable length bytes
+   * format. Keeping this a mix of alphanumeric with special character to avoid
+   * collisions with the regular int/string dictionary values written in fixed size
+   * format.
+   */
+  private static final byte[] MAGIC_HEADER = StringUtil.encodeUtf8("vl1;");
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: What happens when we get to vl10 or come with some other encoding schema say foo1. We would want to keep this backward compatible rt?

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    int length;
+    if (index == getNumElements() - 1) {
+      // This is the last byte array in the buffer so use buffer size to get the length.
+      length = (int) (_dataBuffer.size() - offset);
+    }
+    else {
+      length = _dataBuffer.getInt(INDEX_ARRAY_START_OFFSET + Integer.BYTES * (index + 1)) - offset;
+    }
+
+    byte[] b = buffer == null ? new byte[length] :
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: what is the purpose of these two lines? recursive ternary operators make it hard to parse the code.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java [hunk 5]
```diff
@@ -185,22 +177,14 @@ public void build()
         Preconditions.checkState(numValues > 0);
         _bytesValueToIndexMap = new Object2IntOpenHashMap<>(numValues);
 
+        byte[][] sortedByteArrays = new byte[sortedBytes.length][];
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java:187]: can we avoid creating this?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java [hunk 6]
```diff
@@ -211,6 +195,42 @@ public void build()
     }
   }
 
+  private void writeBytesValueDictionary(int numValues, byte[][] sortedByteArrays) throws IOException {
+    // TODO: Fix this to be dynamic based on the standard deviation in the lengths
+    // of the values in byte array, max length of array, etc.
+    boolean useFixedSizeDict = false;
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java]: make this configurable for now and set the default value to true.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    long length = INDEX_ARRAY_START_OFFSET + Integer.BYTES * (byteArrays.length);
+
+    for (byte[] array: byteArrays) {
+      length += array.length;
+    }
+    return length;
+  }
+
+  public static boolean isVarLengthBytesDictBuffer(PinotDataBuffer buffer) {
+    byte[] header = new byte[MAGIC_HEADER_LENGTH];
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: we need to make this check stronger. There are a few cases this can fail
- dictionary has only one value and its value length is less than vl1;
- if the dictionary has a value vl1; and it happens to be  the first value after sorting the values

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 2 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:48]: Well, the `init()` method is actually writing the data into the buffer, hence I've kept the name ReaderWriter.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 3 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: I've been thinking about this further too. I'll check how this is being done (if at all) in other parts of Pinot so that I can keep it consistent. 

Here are some rough starter ideas I've:
* Keep the length at 8 bytes leaving the scope for future versions and some other bits as well. That will also allow us to add a couple of other char sequences, making the header (almost) uniquely identifiable.
* Keep the size of the header also variable and add the length first. We'll read the length first and then read those many bytes to get the header. The chances of both the length and header colliding with a fixed size dictionary are very very rare.

Do you have any ideas?

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 4 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: The callers of this method might pass a null byte buffer too and we create a new buffer here in that case. The caller knows the buffer length in advance for fixed length dictionary hence it can create and pass the buffer but it's not possible for var length dict, hence we create the buffer here.

I'll simplify the code for readability.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java (hunk 5 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java:187]: We can directly use the `ByteArray[]` but that would duplicate the code in `writeBytesValueDictionary()`. Please note we don't recreate the byte arrays here, just an array of references.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java (hunk 6 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java]: Will do.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 7 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: Good points. Will fix the first one and let's discuss more on second point.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 4 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: This is done.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java (hunk 6 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java]: Done. I've modified the config classes to make this configurable and keep the featured disabled by default.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 7 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: Fixed.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java (hunk 5 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentDictionaryCreator.java:187]: Discussed this with Kishore and explain that we don't recreate the byte arrays, just a new array with references to the same `byte[]` instances.

**Comment by buchireddy** [comment]
FTR, I've done some benchmarking of the new variable length dictionary for strings and compared the numbers with the current FixedLength dictionary and here are the results.

**Time taken to lookup 10M values:**
TimeVsDictionarySizesCharts
**Observation:** As can be clearly seen in the graph, as strings are getting bigger, the variable lengh dictionary is giving much better lookup latencies. When the dictionary size (cardinality) is >1M and the string sizes are small (<100), FixedLength dictionary has better lookup latencies though. 

**Storage requirements of VarLength dictionary:**
Since the variable length dictionary doesn't do any padding, it saves the space for all the cases where the strings in the dictionary aren't of equal length. Hence, this graph plots the % storage savings with VarLength dictionary instead of absolute values.
DictSizeVsStorageSavingsChart
**Observation:** If the strings in the dictionary are of different lengths, VarLength dictionary saves 40% space compared to the fixed length dictionary. 

P.S: All raw values from the benchmarking are available at https://docs.google.com/spreadsheets/d/1iOLyhD4AUZw3JsdOkmH6h36KWYalVeUBIVcty6Pnv0E/edit?usp=sharing so feel free to copy/comment on the results.

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java [hunk 8]
```diff
@@ -275,6 +280,14 @@ public void setColumnSortOrder(List<String> sortOrder) {
     _columnSortOrder.addAll(sortOrder);
   }
 
+  public List<String> getVarLengthDictionaryColumns() {
+    return _varLengthDictionaryColumns;
+  }
+
+  public void setVarLengthDictionaryColumns(List<String> _varLengthDictionaryColumns) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java]: remove _ for param name

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        // Also verify that there is a valid numElements value.
+        int numElements = buffer.getInt(MAGIC_HEADER_LENGTH);
+        return numElements > 0;
+      }
+    }
+
+    return false;
+  }
+
+  public VarLengthBytesValueReaderWriter(PinotDataBuffer dataBuffer) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: move this constructor to the top.

**Review by kishoreg — APPROVED** [review]
Nicely done!.

File: pinot-core/src/main/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java (hunk 8 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java]: Done.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 9 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java]: Done.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+   */
+  private static final byte[] MAGIC_HEADER = StringUtil.encodeUtf8("....vl1;");
+  private static final int MAGIC_HEADER_LENGTH = MAGIC_HEADER.length;
+  private static final int INDEX_ARRAY_START_OFFSET = MAGIC_HEADER.length + Integer.BYTES;
+
+  private final PinotDataBuffer _dataBuffer;
+  private transient int _numElements;
+
+  public VarLengthBytesValueReaderWriter(PinotDataBuffer dataBuffer) {
+    this._dataBuffer = dataBuffer;
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:64]: With '_', 'this' is redundant.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+      length = (int) (_dataBuffer.size() - offset);
+    }
+    else {
+      length = _dataBuffer.getInt(INDEX_ARRAY_START_OFFSET + Integer.BYTES * (index + 1)) - offset;
+    }
+
+    byte[] b;
+    // If the caller didn't pass a buffer, create one with exact length.
+    if (buffer == null) {
+      b = new byte[length];
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:182]: Would thread-local (reusable) byte array help?

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 10 above)
Comment (mayankshriv) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:64]: I think the convention is to use ReaderWriter for mutable case (realtime indices/dictionaries) and Reader for immutable. Will this be used for mutable case as well?

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+ * <ul>
+ *   <li> Integer offsets to start position of byte arrays. </li>
+ *   <li> All byte arrays. </li>
+ * </ul>
+ *
+ * Only sequential writes are supported.
+ *
+ * @see FixedByteValueReaderWriter
+ */
+public class VarLengthBytesValueReaderWriter implements Closeable, ValueReader {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:48]: I am sure there's a reason why you were not able to reuse VarByteChunkSingleValueReader/Writer. Would be good to document that, and may be think if the two can be combined to share some code.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/ImmutableDictionaryReader.java [hunk 13]
```diff
@@ -244,6 +252,6 @@ protected String getPaddedString(int dictId, byte[] buffer) {
   }
 
   protected byte[] getBuffer() {
-    return new byte[_numBytesPerValue];
+    return (_valueReader instanceof VarLengthBytesValueReaderWriter) ? null : new byte[_numBytesPerValue];
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/ImmutableDictionaryReader.java:255]: Is there a way to avoid 'instanceof'?

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public double getDouble(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public String getUnpaddedString(int index, int numBytesPerValue, byte paddingByte, byte[] buffer) {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:154]: Both getUnpaddedString() and getPaddedString() have the same code, and seem to return the same value. The callers of these api's might be expected different values.

**Review by mayankshriv — COMMENTED** [review]
Thanks for doing this, it is a good change. Please consider merging into apache-pinot repo.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 10 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:64]: I agree. Kishore and I have discussed this as well. Since `FixedByteValueReaderWriter` has ReaderWriter in the name, I've named the new class also to keep them consistent. I'll try to rename both in a separate PR.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 11 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:182]: We should be. Let me do that change in another PR.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/ImmutableDictionaryReader.java (hunk 13 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/ImmutableDictionaryReader.java:255]: If we want to avoid it, it's easily doable. Given that `_valueReader` is a final field, we can have a boolean for this condition and use that here.

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 14 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:154]: Correct and IMO, I think the `ValueReader` interface should only have `getString()` rather than padded and unpadded versions but because it has both of those methods, we can't pick/implement only one in VarLengthDictionary. 
Again, let me think more about a better way to clean this up. I wanted to keep the changes to minimal in my very first PR :)

File: pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java (hunk 12 above)
Comment (buchireddy) [reply pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java:48]: Will add the documentation in the next PR.
